cmd/snap/cmd_run.go

```diff
@@ -493,30 +483,90 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+func isPossibleRaceCondition(oldInfo *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock) bool {
+	if !features.RefreshAppAwareness.IsEnabled() || app.IsService() || hintFlock != nil {
+		// Skip check
+		return false
+	}
+
+	currentInfo, err := snap.ReadCurrentInfo(oldInfo.InstanceName())
+	if err != nil {
+		// Current symlink does not exist, snap is either removed or being refreshed.
+		return true
+	}
+	if currentInfo.SnapRevision() != oldInfo.SnapRevision() {
+		// Snap was refreshed while we did not hold the hint lock.
+		return true
+	}
+
+	return false
+}
+
 func (x *cmdRun) snapRunApp(snapApp string, args []string) error {
 	if x.DebugLog {
 		os.Setenv("SNAPD_DEBUG", "1")
 		logger.Debugf("enabled debug logging of early snap startup")
 	}
 	snapName, appName := snap.SplitSnapApp(snapApp)
-	info, err := getSnapInfo(snapName, snap.R(0))
-	if err != nil {
-		return err
-	}
 
-	app := info.Apps[appName]
-	if app == nil {
-		return fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
-	}
+	var retryCnt int
+	for {
+		if retryCnt > 1 {
```

Comment:
Per discussion after the standup. Could we explore the possibility of creating the cgroup earlier in the code flow? The retry might then be unnecessary and perhaps it would make reasoning about the interaction easier if we can keep the same order of locks/checks as snapd.

Specifically, I think we don't need to have exact information obtained by parsing snap.yaml to create a cgroup. If the snap is a service, it will be started by systemd and will already be placed in a service specific cgroup (named. snap.<snap>.<svc>.service). For the app case, we can simply split the snapApp like we do, and create the cgroup right away. Once it's done, snapd will not proceed with the refresh. What we need to do next, is likely to reaffirm that the information we created the cgroup on is correct. i.e. snap name and app name match what is in snap.yaml, and the app name indeed corresponds to an app (or to a service).

Reply:
I really like this, I was worried at first that it might accidentally cause a deadlock somehow with ongoing auto refreshes, but I thought deeply about it and it looks safe and on the upside it would dramatically simplify the flow, no need for the retry logic and no need to hold the hint file lock since no refreshes can start anyway once we create the tracking cgroup.
We might end up reverting https://github.com/snapcore/snapd/pull/13468 and https://github.com/snapcore/snapd/pull/13435 or at least simplifying them if we go with approach.

 I would love to get @pedronis opinion on this approach before attempting a rewrite.

Reply:
I'm thinking about this too, my approval below is for the current approach as-is.

Reply:
After discussing some more with @ZeyadYasser, I think this is probably a better approach. It would certainly be simpler, and I think it's no worse in terms of susceptibility for abuse.

I can't remember exactly what the interaction is between cgroups and refreshes... does the creation of a cgroup abort an existing refresh for the snap, or only prevent a new refresh from beginning?

Currently, what happens in relevant scenarios (to the best of my knowledge):

1. Snap is already running, then refresh is attempted
  - The refresh checks for a cgroup for the running snap
  - The cgroup exists, so the refresh fails, the snap is refresh inhibited, etc.
2. Refresh is beginning (i.e. hasn't yet acquired the inhibit flock), then `snap run` begins
  - The `snap run` does not see the inhibit flock
  - `snap run` creates the cgroup
  - The refresh aborts because of the cgroup
  - The snap runs
3. Refresh is already in progress (holds inhibit flock), then `snap run` begins
  - `snap run` sees that the inhibit lock is held
  - No cgroup is created
  - `snap run` does not start the snap
  - The refresh finishes

### If cgroup creation aborts an in-progress refresh:

If the cgroup were created early, the following would occur in the same scenarios (again, to the best of my knowledge):

1. Snap is already running, then refresh is attempted
  - The refresh checks for a cgroup for the running snap
  - The cgroup exists, so the refresh fails, the snap is refresh inhibited, etc.
2. Refresh is beginning (i.e. hasn't yet acquired the inhibit flock), then `snap run` begins
  - The `snap run` creates the cgroup
  - The refresh aborts because of the cgroup
    - What is the order of checking the cgroup and acquiring the inhibit flock??
  - The snap runs
3. Refresh is already in progress (holds inhibit flock), then `snap run` begins
  - The `snap run` creates the cgroup
  - The refresh aborts because of the cgroup
    - Just to confirm, is this really what happens when the cgroup is created mid-refresh?
  - The snap runs

So, in summary, creating a cgroup early essentially reduces scenarios 2 and 3 to the same result: the refresh is aborted, and the snap runs. This is a more consistent result (no race condition with where in the refresh process the `snap run` occurs), but it would result in notably different behavior from the status quo, if what I said above is correct, and creating a cgroup for a snap causes a refresh of that snap to be aborted.

It would seem that this would eliminate the need for the concept of snap run inhibition in the first place, since `snap run` would always halt an existing refresh. However, this would also mean that refreshes could be cancelled by a malicious program running the snap whenever a refresh is beginning, and thus prevents the snap from ever being refreshed. Perhaps this is already the case, though, since creating the cgroup is unprivileged (afaik), and it's the cgroup which halts the refresh.


### If creating a cgroup only prevents future refreshes from beginning, but does not abort existing refreshes:

If the cgroup were created early, the following would occur in the same scenarios (again, to the best of my knowledge):

1. Snap is already running, then refresh is attempted
  - The refresh checks for a cgroup for the running snap
  - The cgroup exists, so the refresh fails, the snap is refresh inhibited, etc.
2. Refresh is beginning (i.e. hasn't yet acquired the inhibit flock), then `snap run` begins
  - The `snap run` creates the cgroup
  - The refresh aborts because of the cgroup (???)
  - The snap runs
3. Refresh is already in progress (holds inhibit flock), then `snap run` begins
  - The `snap run` creates the cgroup
  - The `snap run` sees that the inhibit flock is held, and aborts
  - The `snap run` deletes the cgroup it just created
    - Does it delete the cgroup here, or wait for the refresh to finish, then run, then delete it?
  - The refresh finishes

So, the difference between this and the status quo is that the transient cgroup is created even if there is an ongoing refresh of the snap, and the cgroup is then deleted after the `snap run` fails. I'm not sure whether this would eliminate the need for snap run retry logic, however, since ongoing refreshes would still prevent the snap from running, and it likely would need to retry once the refresh is completed.

#### begin "What I think is most likely"

If, however, `snap run` does not immediately delete the cgroup if the inhibit flock is locked due to a refresh, then the fact that the cgroup has already been created means another refresh cannot occur until after `snap run` succeeds. This would be very desirable.

The only problem here would be that a malicious unconfined non-root application could perpetually interrupt a refresh from occurring by waiting for a refresh to begin and then running `snap run`, which creates the cgroup early and prevents the refresh from continuing. But since cgroup creation is (afaiu) unprivileged, this is already a concern, and having `snap run` create the cgroup is no more dangerous than an application creating the cgroup directly.

#### end

Some parts of what I said are necessarily contradictory (sorry, I was diverted to other things between talking to @ZeyadYasser and actually writing up these thoughts), so please correct me if I'm missing some obvious points.

Regardless, I think creating a cgroup early would be desirable and decrease complexity, and, as I said, I don't think the risk of using `snap run` to prevent refreshes via early cgroup creation is any worse than the status quo.

Reply:
@bboozzoo @olivercalder @ZeyadYasser 
We want to be able to take the inhibit in some scenarios earlier than it is taken now,  for example if the app itself declare that it is ready to be refreshed, but as people remarked if we move that cgroup creation the purpose of the lock diminishes or gets distorted. So I'm not sure we can do that without risking not being able to take the lock earlier meaningfully.